### ISSUE TITLE
Regarding Tiebreaks

### ISSUE THREAD

**Issue by glasdfghjkl** [Escanord/Intersection_Features#1]
Hi, 

Thank you for your interesting work.

I noticed that in the Test.h file, candidates with the same score don't seem to be accounted for. Could you please let me know where I can find the tiebreak policy for your code?

If there is no tiebreak policy, it might make the comparison somewhat unfair, as typical knowledge graph methods usually break ties by either 1) random selection or 2) using the mean. For instance, if there are 100 candidates with the same score, usual knowledge graph methods would typically either 1) randomly select one from 1 to 100, or 2) assign rank 50.

Thanks in advance for your clarification.

**Comment by glasdfghjkl** [comment]
Dear authors, 

I've ran the code with and without tiebreaks, and the results are quite surprising.
However, I am unsure whether I have executed the code correctly.
I would appreciate it if you could review my modifications and point out any mistakes I may have made.
I have only modified two parts of the code: Test.h->testHead and Test.h->testTail.
Below are the details of my changes:
1. Modified Version of testHead
extern "C"
void testHead(REAL *con, INT lastHead, bool type_constrain = false) {
    INT h = testList[lastHead].h;
    INT t = testList[lastHead].t;
    INT r = testList[lastHead].r;
    INT lef, rig;
    if (type_constrain) {
        lef = head_lef[r];
        rig = head_rig[r];
    }
    REAL minimal = con[h];
    INT l_s = 0;
    INT l_tie = 0;
    INT l_filter_s = 0;
    INT l_filter_tie = 0;
    INT l_s_constrain = 0;
    INT l_filter_s_constrain = 0;

    for (INT j = 0; j < entityTotal; j++) {
        if (j != h) {
            REAL value = con[j];
            if (value < minimal) {
                l_s += 1;
                if (not _find(j, t, r))
                    l_filter_s += 1;
            }
            if (value == minimal) {
                l_tie += 1;
                if (l_tie % 2 == 0){
                    l_s += 1;
                }
                if (not _find(j,t,r)){
                    l_filter_tie += 1;
                    if (l_filter_tie % 2 == 0)
                        l_filter_s += 1;
                }
            }
            if (type_constrain) {
                while (lef < rig && head_type[lef] < j) lef ++;
                if (lef < rig && j == head_type[lef]) {
                    if (value < minimal) {
                        l_s_constrain += 1;
                        if (not _find(j, t, r)) {
                            l_filter_s_constrain += 1;
                        }
                    }  
                }
            }
        }
    }

    if (l_filter_s < 10) l_filter_tot += 1;
    if (l_s < 10) l_tot += 1;
    if (l_filter_s < 3) l3_filter_tot += 1;
    if (l_s < 3) l3_tot += 1;
    if (l_filter_s < 1) l1_filter_tot += 1;
    if (l_s < 1) l1_tot += 1;

    l_filter_rank += (l_filter_s+1);
    l_rank += (1 + l_s);
    l_filter_reci_rank += 1.0/(l_filter_s+1);
    l_reci_rank += 1.0/(l_s+1);

    if (type_constrain) {
        if (l_filter_s_constrain < 10) l_filter_tot_constrain += 1;
        if (l_s_constrain < 10) l_tot_constrain += 1;
        if (l_filter_s_constrain < 3) l3_filter_tot_constrain += 1;
        if (l_s_constrain < 3) l3_tot_constrain += 1;
        if (l_filter_s_constrain < 1) l1_filter_tot_constrain += 1;
        if (l_s_constrain < 1) l1_tot_constrain += 1;

        l_filter_rank_constrain += (l_filter_s_constrain+1);
        l_rank_constrain += (1+l_s_constrain);
        l_filter_reci_rank_constrain += 1.0/(l_filter_s_constrain+1);
        l_reci_rank_constrain += 1.0/(l_s_constrain+1);
    }
}

2. Modified Version of testTail:
extern "C"
void testTail(REAL *con, INT lastTail, bool type_constrain = false) {
    INT h = testList[lastTail].h;
    INT t = testList[lastTail].t;
    INT r = testList[lastTail].r;
    INT lef, rig;
    if (type_constrain) {
        lef = tail_lef[r];
        rig = tail_rig[r];
    }
    REAL minimal = con[t];
    INT r_s = 0;
    INT r_tie = 0;
    INT r_filter_s = 0;
    INT r_filter_tie = 0;
    INT r_s_constrain = 0;
    INT r_filter_s_constrain = 0;
    for (INT j = 0; j < entityTotal; j++) {
        if (j != t) {
            REAL value = con[j];
            if (value < minimal) {
                r_s += 1;
                if (not _find(h, j, r))
                    r_filter_s += 1;
            }
            if (value == minimal) {
                r_tie += 1;
                if (r_tie % 2 == 0){
                    r_s += 1;
                }
                if (not _find(h, j, r)){
                    r_filter_tie += 1;
                    if (r_filter_tie % 2 == 0)
                        r_filter_s += 1;
                }
            }
            if (type_constrain) {
                while (lef < rig && tail_type[lef] < j) lef ++;
                if (lef < rig && j == tail_type[lef]) {
                        if (value < minimal) {
                            r_s_constrain += 1;
                            if (not _find(h, j ,r)) {
                                r_filter_s_constrain += 1;
                            }
                        }
                }
            }
        }
        
    }

    if (r_filter_s < 10) r_filter_tot += 1;
    if (r_s < 10) r_tot += 1;
    if (r_filter_s < 3) r3_filter_tot += 1;
    if (r_s < 3) r3_tot += 1;
    if (r_filter_s < 1) r1_filter_tot += 1;
    if (r_s < 1) r1_tot += 1;

    r_filter_rank += (1+r_filter_s);
    r_rank += (1+r_s);
    r_filter_reci_rank += 1.0/(1+r_filter_s);
    r_reci_rank += 1.0/(1+r_s);
    
    if (type_constrain) {
        if (r_filter_s_constrain < 10) r_filter_tot_constrain += 1;
        if (r_s_constrain < 10) r_tot_constrain += 1;
        if (r_filter_s_constrain < 3) r3_filter_tot_constrain += 1;
        if (r_s_constrain < 3) r3_tot_constrain += 1;
        if (r_filter_s_constrain < 1) r1_filter_tot_constrain += 1;
        if (r_s_constrain < 1) r1_tot_constrain += 1;

        r_filter_rank_constrain += (1+r_filter_s_constrain);
        r_rank_constrain += (1+r_s_constrain);
        r_filter_reci_rank_constrain += 1.0/(1+r_filter_s_constrain);
        r_reci_rank_constrain += 1.0/(1+r_s_constrain);
    }
}

With the tiebreaks, the results are as follows:
metric:			 MRR 		 MR 		 hit@10 	 hit@3  	 hit@1 
averaged(raw):		 0.000157 	 18336.652344 	 0.000000 	 0.000000 	 0.000000 

averaged(filter):	 0.000157 	 18329.187500 	 0.000000 	 0.000000 	 0.000000

Without tiebreaks, the results are:
metric:			 MRR 		 MR 		 hit@10 	 hit@3  	 hit@1 
averaged(raw):		 0.898251 	 367.174530 	 0.899489 	 0.898532 	 0.897735 

averaged(filter):	 0.898403 	 367.155884 	 0.899489 	 0.899011 	 0.897735 

It appears that the results reported in the paper might be due to identical scores for all candidates.

Could you please review my modifications and let me know if I made any errors or if I misunderstood something?
Thank you very much for your time and assistance.

Best regards,

**Comment by albernar** [comment]
I agree with @glasdfghjkl , results are heavily influenced by the absence of a tie-breaking rule.

I think an even better rule is the "pessimistic" one, which assigns, in case of ties, the worst (i.e., highest) possible rank to the target entity. Indeed, in case of a scoring function that squashes values a lot, assigning the mean rank could still give a significant advantage.

The implementation of the "pessimistic" tie-breaking rule simply modifies line 98 and 162 of `openke/base/Test.h` replacing 
`if (value < minimal) {` with `if (value <= minimal) {` in both cases.

The results obtained in that way are as follows:
| Metric               | MRR      | MR          | hit@10  | hit@3   | hit@1   |
|-----------------------|----------|-------------|---------|---------|---------|
| **averaged (raw)**   | 0.000747 | 34310.859375 | 0.003191 | 0.000638 | 0.000000 |
| **averaged (filter)** | 0.000885 | 34295.621094 | 0.003191 | 0.001276 | 0.000000 |

**Comment by Nzteb** [comment]
Hi, do you have any news on this topic?
